### ISSUE TITLE
unrelated_type_equality_checks false positive in NNBD

### ISSUE THREAD

**Issue by a14n** [dart-lang/linter#2196]
In the following code `p` field is nullable but `p` param is non-nullable.

```dart
// @dart = 2.9
class A {
  int? p;
  bool m(num p) => p == this.p; // false positive here
}
```

If I make the `p` field non-nullable the lint disappear.

**Comment by srawlins** [comment]
The case here is that a `int?` is being compared with a `num`. The LUB of `int` and `num` is `num`, and the LUB of `int?` and `num?` is `num?`, but for `int?` and `num`, one allows null and the other does not, so their LUB is `Object?`.

I think we basically need to take nullability out of the equation when deciding whether they are related.
